deltachat/deltachat-android
# Target SDK/API 30

Closes #1849.

Probably best reviewed commit-by-commit and with white-space changes ignored.

- [x] Use the MediaStore API (Scoped Storage) for attachments
- [x] Doing "Add attachment -> Camera" doesn't work yet: https://stackoverflow.com/questions/64392621/android-camera-intent-not-responding-in-android-11
  To fix this, I had to add a `queries` tag to AndroidManifest.xml. But older versions of Gradle don't support this yet, so I also had to update Gradle. I then also had to update Android Studio as Gradle wouldn't work with my old version anymore. And apparently they made some warning a hard error, so I had to add `tools:ignore="AppLinkUrlError"` four times to AndroidManifest.xml (no the warning was not legit, we did everything fine)
- [x] Importing backups (exporting seems to have worked fine before already)
  Unfortunately, I did this by copying the backup file to the private storage first, making importing a backup take longer and require more space unnecessarily.
  The problem is that native code [can only access files that we created ourselves or that are in our private directory](https://developer.android.google.cn/training/data-storage/shared/media?hl=en#direct-file-paths).
  What we have is a `content://` URI, and in Java code we can use it to get an InputStream using `getContentResolver().openInputStream(uri))`.
  Some ideas what we could do to fix this, but none of them sounds too great:
  - Transfer a Stream between the Java and Rust, e.g. via HTTP
  - Hand the Stream object to the Rust code and try to somehow use it
  - Poke around a bit more and see if we find some possibility to still access the file from native code. Mabye I understood the documentation wrongly (it's formulated wishy-washy) and there is a possibility. Like, somehow get the file name from the `content://` URI and then access is from the native code.
  - Re-implement importing backups in Java
- [ ] Lots of testing on lots of Android versions
- [ ] Discuss about the tradeoffs I explained above esp. under "Importing backups"

**Testing checklist**

- Saving a logfile (and check that it was actually saved to "Downloads") (from Settings -> Advanced -> View Log -> Menu -> Save Log)
- Share a file to DC
- Share a file from DC to another app
- Select a file via "Add Attachment -> File"
- Tap "Add Attachment" and check that recent photos can be directly attached from there
- Tap on a file in DC to open it in another app (this only works if the file is NOT an image, video or audio)
- Tap "Add Attachment -> Camera" to take an image using an external camera app
- Export an attachment Check that the shown toast (`File saved to "<folder name>"`) is correct. Check that it was actually saved to this folder.
- Export a backup
- Delete DC, install it again, import the backup. Alternatively, import a backup that was transferred from another device.
- Check that requesting storage permissions works
